### PR TITLE
01 process

### PR DESCRIPTION
**Pull request by martinspetlik**

### CODE REVIEW

File: src/mlmc/tool/flow_mc.py [hunk 1]
```diff
@@ -67,181 +77,247 @@ class FlowSim(simulation.Simulation):
...
+
+    def _make_mesh(self, geo_file, mesh_file, fine_step):
         """
         Make the mesh, mesh_file: <geo_base>_step.msh.
         Make substituted yaml: <yaml_base>_step.yaml,
         using common fields_step.msh file for generated fields.
         :return:
         """
-        subprocess.call([self.env['gmsh'], "-2", '-clscale', str(self.step), '-o', mesh_file, geo_file])
+        subprocess.call([self.env['gmsh'], "-2", '-clscale', str(fine_step), '-o', mesh_file, geo_file])
```
Comment (Paulie14) [inline src/mlmc/tool/flow_mc.py]: I had to add ` '-format', 'msh2',`
```subprocess.call([self.env['gmsh'], "-2", '-format', 'msh2', '-clscale', str(fine_step), '-o', mesh_file, geo_file])```
to get gmsh mesh version 2

**Comment by Paulie14** [comment]
Tried to run the test, but it fails due to incorrect path when writing to a file.
My workdir is `sample`, it creates `sample/output/jobs` dirs but then it tries to create a file
at `sample/output` and fails..

```
Traceback (most recent call last):
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/test/01_cond_field/process.py", line 238, in <module>
    pr = CondField()
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/test/01_cond_field/process.py", line 34, in __init__
    self.run()
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/test/01_cond_field/process.py", line 60, in run
    self.generate_jobs(sampler, n_samples=[100, 50], renew=renew)
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/test/01_cond_field/process.py", line 187, in generate_jobs
    sampler.schedule_samples()
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/src/mlmc/sampler.py", line 134, in schedule_samples
    self._sampling_pool.schedule_sample(sample_id, level_sim)
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/src/mlmc/sampling_pool_pbs.py", line 177, in schedule_sample
    self.serialize_level_sim(level_sim)
  File "/home/paulie/docker_flow123d/WGC2020-THM-MC/MLMC/src/mlmc/sampling_pool_pbs.py", line 196, in serialize_level_sim
    with open(file_path, "wb") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'sample/output/level_0_simulation_config'
```

File: src/mlmc/tool/flow_mc.py [hunk 2]
```diff
@@ -67,181 +77,251 @@ class FlowSim(simulation.Simulation):
...
+        return LevelSimulation(config_dict=config,
+                               task_size=len(fine_mesh_data['points']),
+                               calculate=FlowSim.calculate,  # method which carries out the calculation, will be called from PBS processs
+                               need_sample_workspace=True  # If True, a sample directory is created
+                               )
 
     @staticmethod
     def calculate(config, seed):
         """
+        Method that actually run the calculation, it's called from mlmc.tool.pbs_job.PbsJob.calculate_samples()
```
Comment (jbrezmorf) [inline src/mlmc/tool/flow_mc.py:189]: Add similar docstring to the SimulationBase class.

File: test/01_cond_field/process.py [hunk 3]
```diff
@@ -1,194 +1,111 @@
...
+from mlmc.sample_storage_hdf import SampleStorageHDF
+from mlmc.sampling_pool import OneProcessPool, ProcessPool, ThreadPool
+from mlmc.sampling_pool_pbs import SamplingPoolPBS
+from mlmc.tool import base_process
+from mlmc.tool.flow_mc import FlowSim
+from mlmc.moments import Legendre
+from mlmc.quantity_estimate import QuantityEstimate
 
 
 class CondField(base_process.Process):
```
Comment (jbrezmorf) [inline test/01_cond_field/process.py:15]: `01_cond_field/process.py` should work as a basic tutorial of using MLMC with Flow123d, to this end I propose
to remove support for testing the different number of levels in single run. That way we can make usage of ProcessBase more straightforward.

File: src/mlmc/tool/flow_mc.py [hunk 4]
```diff
@@ -1,19 +1,29 @@
 import os
```
Comment (jbrezmorf) [inline src/mlmc/tool/flow_mc.py:1]: We should refactor the Flow base simulation, but better do it as a next step.

File: src/mlmc/tool/base_process.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+    #     cl = CompareLevels(mlmc_est_list,
+    #                        output_dir=src_path,
+    #                        quantity_name="Q [m/s]",
+    #                        moment_class=Legendre,
+    #                        log_scale=False,
+    #                        n_moments=21, )
+    #
+    #     self.process_analysis(cl)
+
+    def set_environment_variables(self):
```
Comment (jbrezmorf) [inline src/mlmc/tool/base_process.py]: Separate finalized and temporary solution. E.g. various analysis and plotting functions should not be part of ProcessBase. 

The purpose of ProcessBase is to process common script arguments and make a backbone of the sampling script.

File: src/mlmc/tool/base_process.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
+import os
```
Comment (jbrezmorf) [inline src/mlmc/tool/base_process.py]: Common naming scheme:

`process_base.py`, `ProcessBase`

**Comment by jbrezmorf** [comment]
Sampler parameter `step_range`
- do not store as attribute use only in constructor
- rename: `level_parameters: List[float]`

File: test/01_cond_field/process.py [hunk 7]
```diff
@@ -1,194 +1,111 @@
...
-            'output_dir': output_dir,
-            'fields': fields,  # correlated_field.FieldSet object
-            'yaml_file': os.path.join(self.work_dir, '01_conductivity.yaml'),  # The template with a mesh and field placeholders
-            'sim_param_range': self.step_range,  # Range of MLMC simulation parametr. Here the mesh step.
-            'geo_file': os.path.join(self.work_dir, 'square_1x1.geo'),  # The file with simulation geometry (independent of the step)
+            'work_dir': self.work_dir,
+            'env': dict(flow123d=self.flow123d, gmsh=self.gmsh, gmsh_version=1),  # The Environment.
+            'yaml_file': os.path.join(self.work_dir, '01_conductivity.yaml'),
+            # The template with a mesh and field placeholders
+            'sim_param_range': step_range,  # Range of MLMC simulation parametr. Here the mesh step.
```
Comment (jbrezmorf) [inline test/01_cond_field/process.py:83]: Is this used?

File: test/01_cond_field/process.py [hunk 8]
```diff
@@ -1,194 +1,111 @@
...
             'field_template': "!FieldElementwise {mesh_data_file: \"$INPUT_DIR$/%s\", field_name: %s}"
         }
 
-        FlowProcSim.total_sim_id = 0
-
-        self.options['output_dir'] = output_dir
-        mlmc_obj = mlmc.mlmc.MLMC(n_levels, FlowProcSim.factory(self.step_range, config=simulation_config, clean=clean),
-                                  self.step_range, self.options)
+        # Create simulation factory
+        simulation_factory = FlowSim(step_range, config=simulation_config, clean=clean)
```
Comment (jbrezmorf) [inline test/01_cond_field/process.py:91]: Do we need to pass step_range to FlowSim, as a factory it should not be the case.

**Pull request merged**